<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Building a Glycopeptide Hypothesis from a FASTA File &#8212; glycresoft  documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../_static/haiku.css?v=e7493be8" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Search and Analysis Tools" href="../search/index.html" />
    <link rel="prev" title="Glycome Network Definitions" href="glycan-networks.html" /> 
  </head><body>
      <div class="header" role="banner"><h1 class="heading"><a href="../../index.html">
          <span>glycresoft  documentation</span></a></h1>
        <h2 class="heading"><span>Building a Glycopeptide Hypothesis from a FASTA File</span></h2>
      </div>
      <div class="topnav" role="navigation" aria-label="top navigation">
      
        <p>
        «&#160;&#160;<a href="glycan-networks.html">Glycome Network Definitions</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="../search/index.html">Search and Analysis Tools</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <section id="building-a-glycopeptide-hypothesis-from-a-fasta-file">
<h1>Building a Glycopeptide Hypothesis from a FASTA File<a class="headerlink" href="#building-a-glycopeptide-hypothesis-from-a-fasta-file" title="Permalink to this heading">¶</a></h1>
<p>The simplest way to build a glycopeptide database is to start with a
list of theoretical glycoproteins in a FASTA file and perform a simple
<em>in-silico</em> digest with one or more proteases, apply a set of <code class="docutils literal notranslate"><span class="pre">modification</span> <span class="pre">rules</span></code>,
and combine the resulting peptides with a glycan hypothesis to produce
glycopeptides.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Glycan compositions read from the associated glycan hypothesis must be
classified as N-glycans, O-glycans, or GAG-Linkers in order for the algorithm
to select the appropriate glycosite assignment algorithm. Glycan compositions
lacking a classification <strong>will not</strong> be considered.</p>
</div>
<p>The build process is computationally intensive, as such this tool will spawn
several worker processes to share the load.</p>
<section id="glycresoft-build-hypothesis-glycopeptide-fa">
<h2>glycresoft build-hypothesis glycopeptide-fa<a class="headerlink" href="#glycresoft-build-hypothesis-glycopeptide-fa" title="Permalink to this heading">¶</a></h2>
<dl class="simple">
<dt>Constructs a glycopeptide hypothesis from a FASTA file of proteins and a</dt><dd><p>collection of glycans.</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>glycresoft<span class="w"> </span>build-hypothesis<span class="w"> </span>glycopeptide-fa<span class="w"> </span><span class="o">[</span>OPTIONS<span class="o">]</span><span class="w"> </span>FASTA_FILE
<span class="w">                                            </span>DATABASE_CONNECTION
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-z">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-peptide-length-range"></span><span class="sig-name descname"><span class="pre">-z</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--peptide-length-range</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;tuple&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-z" title="Permalink to this definition">¶</a></dt>
<dd><p>The minimum and maximum peptide length to consider  [default: 5, 60]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-g">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-glycan-source"></span><span class="sig-name descname"><span class="pre">-g</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--glycan-source</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-g" title="Permalink to this definition">¶</a></dt>
<dd><p>The path, identity, or other specifier for the glycan source  [required]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-s">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-glycan-source-type"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--glycan-source-type</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;choice&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-s" title="Permalink to this definition">¶</a></dt>
<dd><p>The type of glycan information source to use  [default: text]</p>
<div class="line-block">
<div class="line">Choices: [</div>
<div class="line-block">
<div class="line">combinatorial; text; hypothesis;</div>
<div class="line">analysis]</div>
</div>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-G">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-glycan-source-identifier"></span><span class="sig-name descname"><span class="pre">-G</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--glycan-source-identifier</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-G" title="Permalink to this definition">¶</a></dt>
<dd><p>The name or id number of the hypothesis or analysis to be used when using those glycan source types.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-p">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-processes"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--processes</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of worker processes to use. Defaults to 4 or the number of CPUs, whichever is lower  [default: 4]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-n">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-name"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--name</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-n" title="Permalink to this definition">¶</a></dt>
<dd><p>The name for the hypothesis to be created</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-u">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-occupied-glycosites"></span><span class="sig-name descname"><span class="pre">-u</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--occupied-glycosites</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-u" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of occupied glycosylation sites permitted.  [default: 1]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-e">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-enzyme"></span><span class="sig-name descname"><span class="pre">-e</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--enzyme</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-e" title="Permalink to this definition">¶</a></dt>
<dd><p>The proteolytic enzyme to use during digestion. May be specified multiple times, generating a co-digestion. May specify an enzyme name or a regular expression describing the cleavage pattern. Recognized enzyme names are: 2-iodobenzoate, alpha-lytic protease, arg-c, asp-n, asp-n_ambic, bnps-skatole, caspase 1, caspase 10, caspase 2, caspase 3, caspase 4, caspase 5, caspase 6, caspase 7, caspase 8, caspase 9, chymotrypsin, chymotrypsin high specificity, chymotrypsin low specificity, clostripain, cnbr, enterokinase, factor xa, formic acid, glu-c, glutamyl endopeptidase, granzyme b, hydroxylamine, iodosobenzoic acid, leukocyte elastase, lys-c, lys-c/p, none, ntcb, pepsin ph1.3, pepsin ph2.0, pepsina, proline endopeptidase, proteinase k, staphylococcal peptidase i, thermolysin, thrombin, trypchymo, trypsin, trypsin/p, v8-de, v8-e  [default: trypsin] (May specify more than once)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-m">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-missed-cleavages"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--missed-cleavages</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-m" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of missed proteolytic cleavage sites permitted  [default: 1]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-c">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-constant-modification"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--constant-modification</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Peptide modification rule which will be applied constantly (May specify more than once)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-v">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-variable-modification"></span><span class="sig-name descname"><span class="pre">-v</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--variable-modification</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Peptide modification rule which will be applied variablely (May specify more than once)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-V">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-max-variable-modifications"></span><span class="sig-name descname"><span class="pre">-V</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--max-variable-modifications</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-V" title="Permalink to this definition">¶</a></dt>
<dd><p>The maximum number of variable modifications that can be applied to a single peptide  [default: 4]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-y">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-semispecific-digest"></span><span class="sig-name descname"><span class="pre">-y</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--semispecific-digest</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-y" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a semispecific enzyme digest permitting one peptide terminal to be non-specific  [default: False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-R">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-reverse"></span><span class="sig-name descname"><span class="pre">-R</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--reverse</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-R" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse protein sequences  [default: False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-F">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-not-full-crossproduct"></span><span class="sig-name descname"><span class="pre">-F</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--not-full-crossproduct</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-F" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not produce full crossproduct. For when the search space is too large to enumerate, store, and load.  [default: False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-retain-all-peptides">
<span class="sig-name descname"><span class="pre">--retain-all-peptides</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-retain-all-peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not require a glycosylation site when saving base peptides  [default: False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-C">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-include-n-x-c-glycosylation"></span><span class="sig-name descname"><span class="pre">-C</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--include-n-x-c-glycosylation</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-C" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to support N-x-C in the N-glycosylation sequon  [default: False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-a">
<span id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-annotation-path"></span><span class="sig-name descname"><span class="pre">-a</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--annotation-path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;path&gt;</span></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-a" title="Permalink to this definition">¶</a></dt>
<dd><p>The path to an annotation XML file from UniProt</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-arg-FASTA_FILE">
<span class="sig-name descname"><span class="pre">FASTA_FILE</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-arg-FASTA_FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument
&lt;path&gt; A file containing protein sequences in FASTA format</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-arg-DATABASE_CONNECTION">
<span class="sig-name descname"><span class="pre">DATABASE_CONNECTION</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-glycresoft-build-hypothesis-glycopeptide-fa-arg-DATABASE_CONNECTION" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument
&lt;string&gt; A connection URI for a database, or a path on the file system</p>
</dd></dl>

</section>
<section id="usage-example">
<h2>Usage Example<a class="headerlink" href="#usage-example" title="Permalink to this heading">¶</a></h2>
<p>Below is a basic example of how to use this tool</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># We&#39;ll build a simple glycopeptide hypothesis from just two glycoproteins, the</span>
<span class="c1"># isoforms of Alpha-1-acid glycoprotein.</span>
$<span class="w"> </span>cat<span class="w"> </span>agp.fa
&gt;sp<span class="p">|</span>P02763<span class="p">|</span>A1AG1_HUMAN<span class="w"> </span>Alpha-1-acid<span class="w"> </span>glycoprotein<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="nv">OS</span><span class="o">=</span>Homo<span class="w"> </span>sapiens<span class="w"> </span><span class="nv">GN</span><span class="o">=</span>ORM1<span class="w"> </span><span class="nv">PE</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="nv">SV</span><span class="o">=</span><span class="m">1</span>
MALSWVLTVLSLLPLLEAQIPLCANLVPVPITNATLDQITGKWFYIASAFRNEEYNKSVQ
EIQATFFYFTPNKTEDTIFLREYQTRQDQCIYNTTYLNVQRENGTISRYVGGQEHFAHLL
ILRDTKTYMLAFDVNDEKNWGLSVYADKPETTKEQLGEFYEALDCLRIPKSDVVYTDWKK
DKCEPLEKQHEKERKQEEGES

&gt;sp<span class="p">|</span>P19652<span class="p">|</span>A1AG2_HUMAN<span class="w"> </span>Alpha-1-acid<span class="w"> </span>glycoprotein<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="nv">OS</span><span class="o">=</span>Homo<span class="w"> </span>sapiens<span class="w"> </span><span class="nv">GN</span><span class="o">=</span>ORM2<span class="w"> </span><span class="nv">PE</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="nv">SV</span><span class="o">=</span><span class="m">2</span>
MALSWVLTVLSLLPLLEAQIPLCANLVPVPITNATLDRITGKWFYIASAFRNEEYNKSVQ
EIQATFFYFTPNKTEDTIFLREYQTRQNQCFYNSSYLNVQRENGTVSRYEGGREHVAHLL
FLRDTKTLMFGSYLDDEKNWGLSFYADKPETTKEQLGEFYEALDCLCIPRSDVMYTDWKK
DKCEPLEKQHEKERKQEEGES

<span class="c1"># This hypothesis will include a combinatorial glycan composition hypothesis</span>
<span class="c1"># defined by these rules</span>
$<span class="w"> </span>cat<span class="w"> </span>combinatorial-rules.txt
Hex<span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="m">12</span>
HexNAc<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="m">10</span>
Fuc<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">5</span>
Neu5Ac<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">4</span>

Fuc<span class="w"> </span>&lt;<span class="w"> </span>HexNAc
HexNAc<span class="w"> </span>&gt;<span class="w"> </span>NeuAc<span class="w"> </span>+<span class="w"> </span><span class="m">1</span>

<span class="c1"># We&#39;ll permit the following additional options:</span>
<span class="c1">#  -u 1 means that only one glycosylation site will ever be occupied</span>
<span class="c1">#  -e trypsin means that the proteins will be *in-silico* digested with trypsin</span>
<span class="c1">#  -m 1 means that we will only permit one missed cleavage by the protease</span>
<span class="c1">#  -c &quot;Carbamidomethyl (C)&quot; means we will have a constant modification &quot;Carbamidomethyl&quot;</span>
<span class="c1">#   on all Cysteine residues</span>
<span class="c1">#  -v &quot;Deamidation (N)&quot; means we will have a variable modification &quot;Deamidation&quot; on any</span>
<span class="c1">#   Asparagine residue</span>
<span class="c1">#  -v &quot;Pyro-glu from Q (Q@N-term)&quot; means that we will have a variable modification &quot;Pyro-glu from Q&quot;</span>
<span class="c1">#   on any Glutamine residue on the N-terminus of the peptide sequence</span>
<span class="c1">#  -p 4 means that this task will spawn four worker processes to share the work between</span>
$<span class="w"> </span>glycresoft<span class="w"> </span>build-hypothesis<span class="w"> </span>glycopeptide-fa<span class="w"> </span>-g<span class="w"> </span>combinatorial-rules.txt<span class="w"> </span>-s<span class="w"> </span>combinatorial<span class="se">\</span>
<span class="w">    </span>-u<span class="w"> </span><span class="m">1</span><span class="w"> </span>-e<span class="w"> </span>trypsin<span class="w"> </span>-m<span class="w"> </span><span class="m">1</span><span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;Carbamidomethyl (C)&quot;</span><span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;Deamidation (N)&quot;</span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="s2">&quot;Pyro-glu from Q (Q@N-term)&quot;</span><span class="w"> </span>-p<span class="w"> </span><span class="m">4</span><span class="w"> </span>-n<span class="w"> </span><span class="s2">&quot;Alpha-1-acid Glycopeptide Hypothesis&quot;</span><span class="se">\</span>
<span class="w">    </span>agp.fa<span class="w"> </span>fasta-glycopeptides.db

<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:08.610441<span class="w"> </span>Begin<span class="w"> </span>Combinatorial<span class="w"> </span>Glycan<span class="w"> </span>Hypothesis<span class="w"> </span>Serializer
<span class="o">{</span><span class="s1">&#39;derivatization&#39;</span>:<span class="w"> </span>None,
<span class="w"> </span><span class="s1">&#39;engine&#39;</span>:<span class="w"> </span>Engine<span class="o">(</span>sqlite:///fasta-glycopeptides.db<span class="o">)</span>,
<span class="w"> </span><span class="s1">&#39;glycan_file&#39;</span>:<span class="w"> </span><span class="s1">&#39;combinatorial-rules.txt&#39;</span>,
<span class="w"> </span><span class="s1">&#39;loader&#39;</span>:<span class="w"> </span>None,
<span class="w"> </span><span class="s1">&#39;reduction&#39;</span>:<span class="w"> </span>None,
<span class="w"> </span><span class="s1">&#39;start_time&#39;</span>:<span class="w"> </span>datetime.datetime<span class="o">(</span><span class="m">2017</span>,<span class="w"> </span><span class="m">8</span>,<span class="w"> </span><span class="m">31</span>,<span class="w"> </span><span class="m">2</span>,<span class="w"> </span><span class="m">50</span>,<span class="w"> </span><span class="m">8</span>,<span class="w"> </span><span class="m">608881</span><span class="o">)</span>,
<span class="w"> </span><span class="s1">&#39;status&#39;</span>:<span class="w"> </span><span class="s1">&#39;started&#39;</span>,
<span class="w"> </span><span class="s1">&#39;transformer&#39;</span>:<span class="w"> </span>None,
<span class="w"> </span><span class="s1">&#39;uuid&#39;</span>:<span class="w"> </span><span class="s1">&#39;93604f0929794e2bab9df23e29118b09&#39;</span><span class="o">}</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:08.628096<span class="w"> </span>Generating<span class="w"> </span>Glycan<span class="w"> </span>Compositions<span class="w"> </span>from<span class="w"> </span>Symbolic<span class="w"> </span>Rules<span class="w"> </span><span class="k">for</span><span class="w"> </span>GlycanHypothesis<span class="o">(</span><span class="nv">id</span><span class="o">=</span><span class="m">1</span>,<span class="w"> </span><span class="nv">name</span><span class="o">=</span>GlycanHypothesis-93604f0929794e2bab9df23e29118b09<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:15.188556<span class="w"> </span><span class="m">1000</span><span class="w"> </span>glycan<span class="w"> </span>compositions<span class="w"> </span>created
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.824145<span class="w"> </span>Generated<span class="w"> </span><span class="m">1900</span><span class="w"> </span>glycan<span class="w"> </span>compositions
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.835196<span class="w"> </span>Hypothesis<span class="w"> </span>Completed
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.835577<span class="w"> </span>End<span class="w"> </span>Combinatorial<span class="w"> </span>Glycan<span class="w"> </span>Hypothesis<span class="w"> </span>Serializer
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.835661<span class="w"> </span>Started<span class="w"> </span>at<span class="w"> </span><span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:08.608881.
Ended<span class="w"> </span>at<span class="w"> </span><span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.835265.
Total<span class="w"> </span><span class="nb">time</span><span class="w"> </span>elapsed:<span class="w"> </span><span class="m">0</span>:00:12.226384
CombinatorialGlycanHypothesisSerializer<span class="w"> </span>completed<span class="w"> </span>successfully.
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.903601<span class="w"> </span>Begin<span class="w"> </span>Multiple<span class="w"> </span>Process<span class="w"> </span>Fasta<span class="w"> </span>Glycopeptide<span class="w"> </span>Hypothesis<span class="w"> </span>Serializer
<span class="o">{</span><span class="s1">&#39;constant_modifications&#39;</span>:<span class="w"> </span><span class="o">[</span>Carbamidomethyl:57.021464<span class="o">]</span>,
<span class="w"> </span><span class="s1">&#39;engine&#39;</span>:<span class="w"> </span>Engine<span class="o">(</span>sqlite:///fasta-glycopeptides.db<span class="o">)</span>,
<span class="w"> </span><span class="s1">&#39;fasta_file&#39;</span>:<span class="w"> </span><span class="s1">&#39;agp.fa&#39;</span>,
<span class="w"> </span><span class="s1">&#39;max_glycosylation_events&#39;</span>:<span class="w"> </span><span class="m">1</span>,
<span class="w"> </span><span class="s1">&#39;max_missed_cleavages&#39;</span>:<span class="w"> </span><span class="m">1</span>,
<span class="w"> </span><span class="s1">&#39;n_processes&#39;</span>:<span class="w"> </span><span class="m">4</span>,
<span class="w"> </span><span class="s1">&#39;protease&#39;</span>:<span class="w"> </span><span class="s1">&#39;trypsin&#39;</span>,
<span class="w"> </span><span class="s1">&#39;start_time&#39;</span>:<span class="w"> </span>datetime.datetime<span class="o">(</span><span class="m">2017</span>,<span class="w"> </span><span class="m">8</span>,<span class="w"> </span><span class="m">31</span>,<span class="w"> </span><span class="m">2</span>,<span class="w"> </span><span class="m">50</span>,<span class="w"> </span><span class="m">20</span>,<span class="w"> </span><span class="m">899793</span><span class="o">)</span>,
<span class="w"> </span><span class="s1">&#39;status&#39;</span>:<span class="w"> </span><span class="s1">&#39;started&#39;</span>,
<span class="w"> </span><span class="s1">&#39;total_glycan_combination_count&#39;</span>:<span class="w"> </span>-1,
<span class="w"> </span><span class="s1">&#39;uuid&#39;</span>:<span class="w"> </span><span class="s1">&#39;1e6d202801ce4417b3b153ac84732401&#39;</span>,
<span class="w"> </span><span class="s1">&#39;variable_modifications&#39;</span>:<span class="w"> </span><span class="o">[</span>Deamidated:0.984016,<span class="w"> </span>Gln-&gt;pyro-Glu:-17.026549<span class="o">]}</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.903709<span class="w"> </span>Extracting<span class="w"> </span>Proteins
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.927056<span class="w"> </span>Digesting<span class="w"> </span>Proteins
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:28.965977<span class="w"> </span><span class="m">205</span><span class="w"> </span>Base<span class="w"> </span>Peptides<span class="w"> </span>Produced
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:28.966163<span class="w"> </span>Begin<span class="w"> </span>Applying<span class="w"> </span>Protein<span class="w"> </span>Annotations
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:29.726805<span class="w"> </span>...<span class="w"> </span>Extracting<span class="w"> </span>Best<span class="w"> </span>Peptides
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:29.753947<span class="w"> </span>...<span class="w"> </span>Building<span class="w"> </span>Mask
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:29.755910<span class="w"> </span>...<span class="w"> </span>Removing<span class="w"> </span>Duplicates
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:29.765368<span class="w"> </span>...<span class="w"> </span>Complete
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:29.768705<span class="w"> </span>Combinating<span class="w"> </span>Glycans
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:30.766838<span class="w"> </span>...<span class="w"> </span>Building<span class="w"> </span>combinations<span class="w"> </span><span class="k">for</span><span class="w"> </span>Hypothesis<span class="w"> </span><span class="m">1</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:33.867166<span class="w"> </span><span class="m">1900</span><span class="w"> </span>Glycan<span class="w"> </span>Combinations<span class="w"> </span>Constructed.
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:33.867282<span class="w"> </span>Building<span class="w"> </span>Glycopeptides
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:33.895503<span class="w"> </span>Begin<span class="w"> </span>Creation.<span class="w"> </span>Dropping<span class="w"> </span>Indices
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:33.939597<span class="w"> </span>...<span class="w"> </span>Processing<span class="w"> </span>Glycan<span class="w"> </span>Combinations<span class="w"> </span><span class="m">0</span>-1900<span class="w"> </span><span class="o">(</span><span class="m">100</span>.00%<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021121<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">0</span>-11<span class="w"> </span><span class="m">4</span>.60%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021361<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">11</span>-22<span class="w"> </span><span class="m">9</span>.21%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021429<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">22</span>-33<span class="w"> </span><span class="m">13</span>.81%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021495<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">33</span>-44<span class="w"> </span><span class="m">18</span>.41%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021560<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">44</span>-55<span class="w"> </span><span class="m">23</span>.01%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021624<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">55</span>-66<span class="w"> </span><span class="m">27</span>.62%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021781<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">66</span>-77<span class="w"> </span><span class="m">32</span>.22%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.021890<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">77</span>-88<span class="w"> </span><span class="m">36</span>.82%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.022035<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">88</span>-99<span class="w"> </span><span class="m">41</span>.42%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.022238<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">99</span>-110<span class="w"> </span><span class="m">46</span>.03%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.340675<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">110</span>-121<span class="w"> </span><span class="m">50</span>.63%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.367891<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">121</span>-132<span class="w"> </span><span class="m">55</span>.23%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.376397<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">132</span>-143<span class="w"> </span><span class="m">59</span>.83%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:34.410928<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">143</span>-154<span class="w"> </span><span class="m">64</span>.44%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:46.488753<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">154</span>-165<span class="w"> </span><span class="m">69</span>.04%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:48.904966<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">165</span>-176<span class="w"> </span><span class="m">73</span>.64%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:52.744931<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">176</span>-187<span class="w"> </span><span class="m">78</span>.24%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:52.765699<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">187</span>-198<span class="w"> </span><span class="m">82</span>.85%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:52.806290<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">198</span>-209<span class="w"> </span><span class="m">87</span>.45%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:59.408814<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">209</span>-220<span class="w"> </span><span class="m">92</span>.05%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:51:08.196088<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">220</span>-231<span class="w"> </span><span class="m">96</span>.65%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:51:11.241885<span class="w"> </span>...<span class="w"> </span>Dealt<span class="w"> </span>Peptides<span class="w"> </span><span class="m">231</span>-239<span class="w"> </span><span class="m">100</span>.00%
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:51:11.241991<span class="w"> </span>...<span class="w"> </span>All<span class="w"> </span>Peptides<span class="w"> </span>Dealt
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:51:33.927773<span class="w"> </span>...<span class="w"> </span><span class="m">130001</span><span class="w"> </span>Glycopeptides<span class="w"> </span>Created
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:10.984681<span class="w"> </span>Process<span class="w"> </span><span class="m">3560</span><span class="w"> </span>completed.<span class="w"> </span><span class="o">(</span><span class="m">41</span><span class="w"> </span>peptides,<span class="w"> </span><span class="m">57000</span><span class="w"> </span>glycopeptides<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:11.928619<span class="w"> </span>Process<span class="w"> </span><span class="m">3556</span><span class="w"> </span>completed.<span class="w"> </span><span class="o">(</span><span class="m">66</span><span class="w"> </span>peptides,<span class="w"> </span><span class="m">55100</span><span class="w"> </span>glycopeptides<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:12.924645<span class="w"> </span>Process<span class="w"> </span><span class="m">3558</span><span class="w"> </span>completed.<span class="w"> </span><span class="o">(</span><span class="m">66</span><span class="w"> </span>peptides,<span class="w"> </span><span class="m">57000</span><span class="w"> </span>glycopeptides<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:17.504664<span class="w"> </span>Process<span class="w"> </span><span class="m">3562</span><span class="w"> </span>completed.<span class="w"> </span><span class="o">(</span><span class="m">66</span><span class="w"> </span>peptides,<span class="w"> </span><span class="m">62700</span><span class="w"> </span>glycopeptides<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:19.507377<span class="w"> </span>Joining<span class="w"> </span>Process<span class="w"> </span><span class="m">3556</span><span class="w"> </span><span class="o">(</span>False<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:19.507669<span class="w"> </span>Joining<span class="w"> </span>Process<span class="w"> </span><span class="m">3558</span><span class="w"> </span><span class="o">(</span>False<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:19.507839<span class="w"> </span>Joining<span class="w"> </span>Process<span class="w"> </span><span class="m">3560</span><span class="w"> </span><span class="o">(</span>False<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:19.507991<span class="w"> </span>Joining<span class="w"> </span>Process<span class="w"> </span><span class="m">3562</span><span class="w"> </span><span class="o">(</span>False<span class="o">)</span>
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:19.508170<span class="w"> </span>All<span class="w"> </span>Work<span class="w"> </span>Done.<span class="w"> </span>Rebuilding<span class="w"> </span>Indices
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.342381<span class="w"> </span>Analyzing<span class="w"> </span>Indices
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.605914<span class="w"> </span>Done<span class="w"> </span>Analyzing<span class="w"> </span>Indices
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.645477<span class="w"> </span>Generated<span class="w"> </span><span class="m">231800</span><span class="w"> </span>glycopeptides
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.645579<span class="w"> </span>Done
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.659734<span class="w"> </span>Hypothesis<span class="w"> </span>Completed
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.660019<span class="w"> </span>End<span class="w"> </span>Multiple<span class="w"> </span>Process<span class="w"> </span>Fasta<span class="w"> </span>Glycopeptide<span class="w"> </span>Hypothesis<span class="w"> </span>Serializer
<span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.660143<span class="w"> </span>Started<span class="w"> </span>at<span class="w"> </span><span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:50:20.899793.
Ended<span class="w"> </span>at<span class="w"> </span><span class="m">2017</span>-08-31<span class="w"> </span><span class="m">02</span>:52:21.659825.
Total<span class="w"> </span><span class="nb">time</span><span class="w"> </span>elapsed:<span class="w"> </span><span class="m">0</span>:02:00.760032
MultipleProcessFastaGlycopeptideHypothesisSerializer<span class="w"> </span>completed<span class="w"> </span>successfully.
</pre></div>
</div>
<p>If you instead wish to use an existing glycan hypothesis instead
of creating a new one, you can modify the instructions above:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>glycresoft<span class="w"> </span>build-hypothesis<span class="w"> </span>glycan-combinatorial<span class="w"> </span>rules-file.txt<span class="w"> </span>combinatorial-database<span class="w"> </span>-n<span class="w"> </span><span class="s2">&quot;Combinatorial Human N-Glycans&quot;</span>
...

$<span class="w"> </span>glycresoft<span class="w"> </span>build-hypothesis<span class="w"> </span>glycopeptide-fa<span class="w"> </span>-g<span class="w"> </span>combinatorial-database.db<span class="w"> </span>-s<span class="w"> </span>hypothesis<span class="w"> </span>-G<span class="w"> </span><span class="m">1</span><span class="se">\</span>
<span class="w">    </span>-u<span class="w"> </span><span class="m">1</span><span class="w"> </span>-e<span class="w"> </span>trypsin<span class="w"> </span>-m<span class="w"> </span><span class="m">1</span><span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;Carbamidomethyl (C)&quot;</span><span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;Deamidation (N)&quot;</span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="s2">&quot;Pyro-glu from Q (Q@N-term)&quot;</span><span class="w"> </span>-p<span class="w"> </span><span class="m">4</span><span class="w"> </span>-n<span class="w"> </span><span class="s2">&quot;Alpha-1-acid Glycopeptide Hypothesis&quot;</span><span class="se">\</span>
<span class="w">    </span>agp.fa<span class="w"> </span>fasta-glycopeptides.db
...
</pre></div>
</div>
<p>The primary difference here is that the value of the <code class="docutils literal notranslate"><span class="pre">-g</span></code> option is the path to the source
database’s path (or connection URI), <code class="docutils literal notranslate"><span class="pre">-s</span></code> indicates that the source is a hypothesis, and
the new <code class="docutils literal notranslate"><span class="pre">-G</span></code> option identifies the hypothesis to use, as a single database may contain many
hypotheses. The value of <code class="docutils literal notranslate"><span class="pre">-G</span></code> can be the ID of the hypothesis or the hypothesis’s name (in this case
“Combinatorial Human N-Glycans”).</p>
<section id="supported-proteases">
<h3>Supported Proteases<a class="headerlink" href="#supported-proteases" title="Permalink to this heading">¶</a></h3>
<div class="admonition error">
<p class="admonition-title">Error</p>
<p>Unable to execute python code at fasta-glycopeptide.rst:189:</p>
<p>object of type ‘map’ has no len()</p>
</div>
</section>
<section id="supported-modification-rules">
<h3>Supported Modification Rules<a class="headerlink" href="#supported-modification-rules" title="Permalink to this heading">¶</a></h3>
<p><cite>glycresoft</cite> supports the full range of <a class="reference external" href="http://www.unimod.org/modifications_list.php?">UNIMOD</a>
modification rules as well as some common alternative namings.</p>
<p>To be more specific, you are able to override modification targets when you
specify modification names by passing the permitted target rules enclosed in
parentheses following the modification name. For example “Deamidation (N)” will
only target Asparagine residues, unlike the plain “Deamidation” rule which will
target both Asparagine and Glutamine.</p>
<p>For more information about supported post-translational modifications, please
see <a class="reference external" href="todo">Peptide Modifications</a></p>
</section>
<section id="uniprot-integration">
<h3>UniProt Integration<a class="headerlink" href="#uniprot-integration" title="Permalink to this heading">¶</a></h3>
<p>todo</p>
</section>
</section>
</section>


      </div>
      <div class="bottomnav" role="navigation" aria-label="bottom navigation">
      
        <p>
        «&#160;&#160;<a href="glycan-networks.html">Glycome Network Definitions</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="../search/index.html">Search and Analysis Tools</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer" role="contentinfo">
    &#169; Copyright 2024, Joshua Klein.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.1.2.
    </div>
  </body>
</html>